### PR TITLE
Contiguous tensor product IndexSet.

### PR DESCRIPTION
**Pull request by luca-heltai**
Followup to #9074

### CODE REVIEW

File: include/deal.II/base/index_set.h [hunk 1]
```diff
@@ -352,6 +352,14 @@ class IndexSet
   void
   subtract_set(const IndexSet &other);
 
+  /**
+   * Return a new IndexSet, with global size equal to `n_comps*this->size()`,
+   * containing for every element `n` of this IndexSet, the half open range
+   * `[n*n_comps, (n+1)*n_comps)`.
+   */
```
Comment (bangerth) [inline include/deal.II/base/index_set.h]:
```suggestion
   *
   * The name results from the perspective that one starts with an IndexSet and takes the tensor
   * product with a "mask" `[1,...,1]` that contains `n_comp` elements; this results in a matrix
   * of size `this->size()` times `n_comps` that has ones in exactly the rows for which the IndexSet
   * contained an index. This matrix is then "unrolled" again by going through each row one by one
   * and reindexing the entries of the matrix in consecutive order. A one in the matrix then corresponds
   * to an entry in the reindexed IndexSet that is returned by this function.
   */
```

**Comment by luca-heltai** [comment]
> OK if you add something like my suggestion to the documentation. I think it would have been interesting to write the function in a way so that it takes another IndexSet or vector<bool> or ComponentMask, etc, as argument. But this will do.

That's simple enough. I'll write it as you suggest. :)

File: doc/news/changes/minor/20191212LucaHeltai [hunk 2]
```diff
@@ -0,0 +1,5 @@
+New: IndexSet::contiguous_tensor_product() creates a new IndexSet with global size equal to
```
Comment (tjhei) [inline doc/news/changes/minor/20191212LucaHeltai]: update this entry please (function name, description)

**Pull request merged**